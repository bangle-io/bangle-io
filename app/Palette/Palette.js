import React, { useContext } from 'react';
import { useCallback } from 'react';
import { UIManagerContext } from 'ui-context/index';

import { useKeybindings } from 'utils/index';

import { PaletteUI } from '../PaletteUI/PaletteUI';
import { useCommandPalette } from './Palettes/CommandPalette';
import { useFilePalette } from './Palettes/FilePalette';
import { useInputPalette } from './Palettes/InputPalette';
import {
  COMMAND_PALETTE,
  FILE_PALETTE,
  INPUT_PALETTE,
  WORKSPACE_PALETTE,
} from './paletteTypes';
import { useWorkspacePalette } from './Palettes/WorkspacePalette';
import { FileDocumentIcon, AlbumIcon, TerminalIcon } from '../helper-ui/Icons';

export function Palette() {
  const {
    paletteMetadata: metadata,
    paletteType,
    paletteInitialQuery,
    dispatch,
    widescreen,
  } = useContext(UIManagerContext);

  const updatePalette = useCallback(
    ({ type, initialQuery, metadata }) => {
      dispatch({
        type: 'UI/CHANGE_PALETTE_TYPE',
        value: { type: type, initialQuery, metadata },
      });
    },
    [dispatch],
  );

  usePaletteKeybindings({ updatePalette });

  const paletteItems = [
    useFilePalette(),
    useWorkspacePalette({ updatePalette }),
    useCommandPalette({ updatePalette }),
    useInputPalette({ metadata, updatePalette }),
  ];

  return (
    <PaletteUI
      paletteTypeIcon={getPaletteIcon(paletteType)}
      paletteType={paletteType}
      updatePalette={updatePalette}
      paletteInitialQuery={paletteInitialQuery}
      parseRawQuery={parseRawQuery}
      generateRawQuery={generateRawQuery}
      paletteItems={paletteItems}
      className={`fadeInScaleAnimation bangle-palette ${
        widescreen ? 'widescreen' : ''
      }`}
    />
  );
}

function usePaletteKeybindings({ updatePalette }) {
  useKeybindings(() => {
    return {
      'Mod-P': () => {
        updatePalette({ type: COMMAND_PALETTE });
        return true;
      },

      'Mod-p': () => {
        updatePalette({ type: FILE_PALETTE });
        return true;
      },

      'Ctrl-r': () => {
        updatePalette({ type: WORKSPACE_PALETTE });
        return true;
      },
    };
  }, [updatePalette]);
}

const parseRawQuery = (currentType, rawQuery) => {
  // Some of the types depend on the current active query
  // for example if query starts with `>`, it becomes a command type
  // and if a user backspaces `>` it defaults to file.
  // but thats not true for all as `command/input/*` is static
  // and can only be dismissed.
  if (rawQuery.startsWith('>')) {
    return { paletteType: COMMAND_PALETTE, query: rawQuery.slice(1) };
  }

  if (rawQuery.startsWith('ws:')) {
    return { paletteType: WORKSPACE_PALETTE, query: rawQuery.slice(3) };
  }

  // Disallow changing of palette type
  if (currentType === INPUT_PALETTE) {
    return { paletteType: currentType, query: rawQuery };
  }

  return { paletteType: FILE_PALETTE, query: rawQuery };
};

const generateRawQuery = (paletteType, query) => {
  if (paletteType === COMMAND_PALETTE) {
    return '>' + query;
  }

  if (paletteType === WORKSPACE_PALETTE) {
    return 'ws:' + query;
  }

  if (paletteType === INPUT_PALETTE) {
    return query;
  }

  // defaults to file
  return query;
};

function getPaletteIcon(paletteType) {
  let Icon;
  switch (paletteType) {
    case FILE_PALETTE: {
      Icon = FileDocumentIcon;
      break;
    }
    case COMMAND_PALETTE: {
      Icon = TerminalIcon;
      break;
    }
    case WORKSPACE_PALETTE: {
      Icon = AlbumIcon;
      break;
    }
    default: {
      return null;
    }
  }

  return (
    <span className="pr-2 flex items-center">
      <Icon className="h-5 w-5 " />
    </span>
  );
}
