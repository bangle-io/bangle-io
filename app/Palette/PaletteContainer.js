import React, { createRef } from 'react';
import { useWatchClickOutside } from 'bangle-io/app/misc/hooks';

import { useFilePalette } from './Palettes/FilePalette';
import { PaletteInput } from './PaletteInput';
import { useWorkspacePalette } from './Palettes/WorkspacePalette';
import { PaletteType } from './PaletteType';

const parseRawQuery = (rawQuery, paletteType) => {
  // Some of the types depend on the current active query
  // for example if query starts with `>`, it becomes a command type
  // and if a user backspaces `>` it defaults to file.
  // but thats not true for all as `command/input/*` is static
  // and can only be dismissed.
  if (rawQuery.startsWith('>')) {
    return { paletteType: 'command', query: rawQuery.slice(1) };
  }

  if (rawQuery.startsWith('ws:')) {
    return { paletteType: 'workspace', query: rawQuery.slice(3) };
  }

  // Disallow changing of palette type
  if (paletteType.startsWith('command/input/')) {
    return { paletteType, query: rawQuery };
  }

  return { paletteType: 'file', query: rawQuery };
};

const generateRawQuery = (paletteType, query) => {
  if (paletteType === 'command') {
    return '>' + query;
  }

  if (paletteType === 'workspace') {
    return 'ws:' + query;
  }

  if (paletteType.startsWith('command/input/')) {
    return query;
  }

  // defaults to file
  return query;
};

export function PaletteContainer({
  updateExecuteActiveItem,
  executeActiveItem,
  updatePaletteType,
  paletteType,
  updateCounter,
  updateQuery,
  counter,
  query,
}) {
  const paletteInputRef = createRef();

  const containerRef = useWatchClickOutside(
    () => {
      updatePaletteType(null);
    },
    () => {
      paletteInputRef.current.focus();
    },
  );
  const filePaletteItems = useFilePalette({ query });
  const workspacePaletteItems = useWorkspacePalette({ query });

  const onDismiss = () => {
    updatePaletteType(null);
  };
  return (
    <div
      className="bangle-palette z-30 p-2 shadow-md border flex flex-col"
      ref={containerRef}
    >
      <PaletteInput
        ref={paletteInputRef}
        onDismiss={onDismiss}
        onPressEnter={() => {
          // To prevent multiple executions
          if (!executeActiveItem) {
            updateExecuteActiveItem(true);
          }
        }}
        updateCounter={updateCounter}
        updateQuery={(rawQuery) => {
          const initialPaletteType = paletteType;
          const { paletteType: newType, query } = parseRawQuery(
            rawQuery,
            initialPaletteType,
          );

          if (newType !== initialPaletteType) {
            updatePaletteType(newType, query);
          } else {
            updateQuery(query);
          }
        }}
        query={generateRawQuery(paletteType, query)}
        counter={counter}
      />
      <PaletteType
        match={/^file$/}
        paletteItems={filePaletteItems}
        executeActiveItem={executeActiveItem}
        counter={counter}
        paletteType={paletteType}
        onDismiss={onDismiss}
      />
      <PaletteType
        match={/^workspace$/}
        paletteItems={workspacePaletteItems}
        executeActiveItem={executeActiveItem}
        counter={counter}
        paletteType={paletteType}
        onDismiss={onDismiss}
      />
      {/* <PaletteItem match={/^(command|command\/input\/.*)$/}>
        <CommandPalette {...props} />
      </PaletteItem>  */}
    </div>
  );
}
