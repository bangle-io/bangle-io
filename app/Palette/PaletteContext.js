import React, { useState, useCallback, useMemo } from 'react';
import { getLast } from '../misc/index';

export const PaletteContext = React.createContext();

export function PaletteContextProvider({ children }) {
  const [state, setState] = useState({ stack: [], execute: false });

  const push = useCallback(
    (val) => {
      setState({
        ...state,
        stack: [val],
        execute: false,
      });
    },
    [setState, state],
  );

  const onExecute = useCallback(
    (val = true) => {
      setState({
        ...state,
        execute: val,
      });
    },
    [setState, state],
  );

  const updateCurrent = useCallback(
    (val) => {
      const current = getLast(state.stack);
      let newCurrent = typeof val === 'function' ? val(current) : val;

      setState({
        ...state,
        execute: false,
        stack: [
          ...state.stack.slice(0, state.stack.length - 1),
          { ...current, ...newCurrent },
        ],
      });
    },
    [setState, state],
  );

  const clear = useCallback(
    (val) => {
      setState({
        ...state,
        stack: [],
        execute: false,
      });
    },
    [setState, state],
  );

  const value = useMemo(() => {
    return {
      updateCurrent,
      push,
      clear,
      execute: state.execute,
      onExecute,
      current: getLast(state.stack),
    };
  }, [clear, updateCurrent, push, state, onExecute]);

  return (
    <PaletteContext.Provider value={value}>{children}</PaletteContext.Provider>
  );
}
