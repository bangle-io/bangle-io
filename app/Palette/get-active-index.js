/**
 * Calculate the currently active item
 * @param {*} counter The currently active counter passed to you by this component
 * @param {*} size The total number of elements displayed after applying query
 */
export const getActiveIndex = (counter, size) => {
  const r = counter % size;
  return r < 0 ? r + size : r;
};
