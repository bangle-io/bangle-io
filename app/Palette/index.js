export * from './PaletteInput';
export * from './Palette';
