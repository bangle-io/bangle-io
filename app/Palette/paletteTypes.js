import {
  AlbumIcon,
  FileDocumentIcon,
  NullIcon,
  TerminalIcon,
} from 'ui-components/index';
import { keybindings } from 'config/index';

export const FILE_PALETTE = 'file';
export const COMMAND_PALETTE = 'command';
export const WORKSPACE_PALETTE = 'workspace';
export const INPUT_PALETTE = 'input';
export const QUESTION_PALETTE = 'question';

export const palettes = {
  [FILE_PALETTE]: {
    type: FILE_PALETTE,
    Icon: FileDocumentIcon,
    keybinding: keybindings.toggleFilePalette.key,
    inputPlaceholder: `Enter a file name or type '?' to see other palettes.`,
  },
  [COMMAND_PALETTE]: {
    type: COMMAND_PALETTE,
    Icon: TerminalIcon,
    keybinding: keybindings.toggleCommandPalette.key,
    inputPlaceholder: null,
  },
  [WORKSPACE_PALETTE]: {
    type: WORKSPACE_PALETTE,
    Icon: AlbumIcon,
    keybinding: keybindings.toggleWorkspacePalette.key,
    inputPlaceholder: null,
  },
  [INPUT_PALETTE]: {
    type: INPUT_PALETTE,
    Icon: NullIcon,
    keybinding: null,
    inputPlaceholder: null,
  },
  [QUESTION_PALETTE]: {
    type: QUESTION_PALETTE,
    Icon: NullIcon,
    keybinding: null,
    inputPlaceholder: null,
  },
};
