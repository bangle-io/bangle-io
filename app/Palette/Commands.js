import { INPUT_PALETTE } from './paletteTypes';
import { useCreateMdFile, useWorkspacePath } from 'workspace/index';
import { useCallback, useContext } from 'react';
import { UIManagerContext } from 'ui-context';

/**
 * Opens an input palette
 */
export function useInputPaletteNewFileCommand() {
  const createNewFile = useCreateMdFile();
  const { wsName } = useWorkspacePath();

  const { dispatch } = useContext(UIManagerContext);

  const createFile = useCallback(
    ({ initialQuery = '' } = {}) => {
      dispatch({
        type: 'UI/CHANGE_PALETTE_TYPE',
        value: {
          type: INPUT_PALETTE,
          initialQuery,
          metadata: {
            onInputConfirm: (query) => {
              let normalizedQuery = query;
              if (!normalizedQuery.endsWith('.md')) {
                normalizedQuery += '.md';
              }
              return createNewFile(wsName + ':' + normalizedQuery);
            },
          },
        },
      });
    },
    [createNewFile, dispatch, wsName],
  );

  return createFile;
}
