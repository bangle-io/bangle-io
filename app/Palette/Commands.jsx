import { INPUT_PALETTE } from './paletteTypes';
import { useCreateNote, useWorkspacePath } from 'workspace/index';
import { useCallback, useContext } from 'react';
import { UIManagerContext } from 'ui-context';
import { EditorManagerContext } from '../editor/EditorManager';

/**
 * Opens an input palette
 */
export function useInputPaletteNewNoteCommand() {
  const { bangleIOContext } = useContext(EditorManagerContext);

  const createNote = useCreateNote();
  const { wsName } = useWorkspacePath();

  const { dispatch } = useContext(UIManagerContext);

  const createNoteCallback = useCallback(
    ({ initialQuery = '' } = {}) => {
      dispatch({
        type: 'UI/CHANGE_PALETTE_TYPE',
        value: {
          type: INPUT_PALETTE,
          initialQuery,
          metadata: {
            paletteInfo: 'You are currently creating a new note',
            placeholder: 'Type the name of the note to create',
            onInputConfirm: (query) => {
              let normalizedQuery = query;
              if (!query) {
                return Promise.reject(new Error('Must provide a note name'));
              }
              if (!normalizedQuery.endsWith('.md')) {
                normalizedQuery += '.md';
              }
              return createNote(
                bangleIOContext,
                wsName + ':' + normalizedQuery,
              );
            },
          },
        },
      });
    },
    [createNote, bangleIOContext, dispatch, wsName],
  );

  return createNoteCallback;
}

/**
 * Returns a callback which will accept a bangle command and ...params
 * and on calling this callback it will execute the command.
 */
export function useDispatchPrimaryEditorCommand(dry = true) {
  const { primaryEditor } = useContext(EditorManagerContext);
  return useCallback(
    (editorCommand, ...params) => {
      if (!primaryEditor || primaryEditor.destroyed) {
        return false;
      }
      const { dispatch, state } = primaryEditor.view;
      if (dry) {
        const result = editorCommand(...params)(state);
        return result;
      }

      return editorCommand(...params)(state, dispatch, primaryEditor.view);
    },
    [dry, primaryEditor],
  );
}
