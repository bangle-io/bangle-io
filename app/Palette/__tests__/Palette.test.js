import React, { useContext } from 'react';
import userEvent from '@testing-library/user-event';
import { render, act } from '@testing-library/react';
import { UIManager, UIManagerContext } from 'bangle-io/app/UIManager';
import {
  useGetWorkspaceFiles,
  useWorkspacePath,
  useWorkspaces,
} from 'bangle-io/app/workspace/workspace-hooks';
import { COMMAND_PALETTE, FILE_PALETTE, INPUT_PALETTE } from '../paletteTypes';
import { Palette } from '../../Palette/Palette';
import { sleep } from 'bangle-io/app/misc/index';

let result, paletteType, paletteInitialQuery, dispatch;
jest.mock('bangle-io/app/workspace/workspace-hooks', () => {
  return {
    useWorkspacePath: jest.fn(),
    useWorkspaces: jest.fn(),
    useGetWorkspaceFiles: jest.fn(),
    useCreateMdFile: jest.fn(),
    useRenameActiveFile: jest.fn(),
    useDeleteFile: jest.fn(),
  };
});

beforeEach(async () => {
  useGetWorkspaceFiles.mockImplementation(jest.fn(() => [[], jest.fn()]));
  useWorkspaces.mockImplementation(jest.fn(() => ({ workspaces: [] })));
  useWorkspacePath.mockImplementation(jest.fn(() => ({})));

  paletteType = undefined;
  paletteInitialQuery = undefined;
  dispatch = null;
  function Comp() {
    ({ paletteType, dispatch } = useContext(UIManagerContext));

    return (
      <div>
        <Palette />
      </div>
    );
  }
  result = await render(
    <UIManager>
      <Comp />
    </UIManager>,
  );
});

test('Empty on mount', async () => {
  expect(result.container).toMatchInlineSnapshot(`
    <div>
      <div />
    </div>
  `);

  expect(paletteType).toBeUndefined();
});

test('Correctly switches to file type', async () => {
  let promise = Promise.resolve();

  useGetWorkspaceFiles.mockImplementation(() => {
    return [['my-ws:one.md'], jest.fn()];
  });
  act(() => {
    dispatch({
      type: 'UI/CHANGE_PALETTE_TYPE',
      value: { type: FILE_PALETTE },
    });
  });

  await act(() => promise);
  await act(() => new Promise((res) => setTimeout(res, 0)));

  expect([...result.container.querySelectorAll('.side-bar-row')])
    .toMatchInlineSnapshot(`
    Array [
      <div
        class="flex side-bar-row flex-row items-center cursor-pointer  active "
        style="padding-left: 16px; padding-right: 16px;"
      >
        <span
          class="text-lg truncate select-none"
        >
          one.md
        </span>
        <span
          class="flex-1 flex "
        />
      </div>,
    ]
  `);

  expect(result.container).toMatchSnapshot();
});

test('Correctly filters commands', async () => {
  act(() => {
    dispatch({
      type: 'UI/CHANGE_PALETTE_TYPE',
      value: { type: FILE_PALETTE },
    });
  });

  const input = result.getByLabelText('palette-input');
  userEvent.type(input, '>');

  expect(result.container.querySelectorAll('.side-bar-row')).toMatchSnapshot();

  userEvent.type(input, 'toggle');

  expect(result.container.querySelectorAll('.side-bar-row'))
    .toMatchInlineSnapshot(`
    NodeList [
      <div
        class="flex side-bar-row flex-row items-center cursor-pointer  active "
        style="padding-left: 16px; padding-right: 16px;"
      >
        <span
          class="text-lg truncate select-none"
        >
          View: Toggle theme
        </span>
        <span
          class="flex-1 flex "
        />
      </div>,
      <div
        class="flex side-bar-row flex-row items-center cursor-pointer   "
        style="padding-left: 16px; padding-right: 16px;"
      >
        <span
          class="text-lg truncate select-none"
        >
          View: Toggle sidebar
        </span>
        <span
          class="flex-1 flex "
        />
      </div>,
    ]
  `);
});

test('input palette', async () => {
  const initialQuery = 'hello';
  const onInputConfirm = jest.fn();
  act(() => {
    dispatch({
      type: 'UI/CHANGE_PALETTE_TYPE',
      value: {
        type: INPUT_PALETTE,
        initialQuery,
        metadata: {
          onInputConfirm,
        },
      },
    });
  });

  const input = result.getByLabelText('palette-input');
  expect(paletteType).toBe(INPUT_PALETTE);

  expect(input.value).toBe(initialQuery);

  await act(async () => {
    userEvent.type(input, '{enter}');
    // to let uiManager settle
    await sleep(0);
  });
  expect(onInputConfirm).toBeCalledTimes(1);
  expect(onInputConfirm).nthCalledWith(1, 'hello');

  expect(paletteType).toBe(null);
});

test.skip('Keybindings work', async () => {
  userEvent.keyboard('{ControlLeft>}r{/ControlLeft}');
  expect(result.container).toMatchInlineSnapshot(`
      <div>
        <div />
      </div>
  `);
});
