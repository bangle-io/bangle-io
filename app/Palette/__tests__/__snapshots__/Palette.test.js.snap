// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly filters commands 1`] = `
NodeList [
  <div
    class="flex side-bar-row flex-row items-center cursor-pointer  active "
    style="padding-left: 16px; padding-right: 16px;"
  >
    <span
      class="text-lg truncate select-none"
    >
      View: Toggle theme
    </span>
    <span
      class="flex-1 flex "
    />
  </div>,
  <div
    class="flex side-bar-row flex-row items-center cursor-pointer   "
    style="padding-left: 16px; padding-right: 16px;"
  >
    <span
      class="text-lg truncate select-none"
    >
      View: Toggle sidebar
    </span>
    <span
      class="flex-1 flex "
    />
  </div>,
  <div
    class="flex side-bar-row flex-row items-center cursor-pointer   "
    style="padding-left: 16px; padding-right: 16px;"
  >
    <span
      class="text-lg truncate select-none"
    >
      Workspace: New File
    </span>
    <span
      class="flex-1 flex "
    />
  </div>,
  <div
    class="flex side-bar-row flex-row items-center cursor-pointer   "
    style="padding-left: 16px; padding-right: 16px;"
  >
    <span
      class="text-lg truncate select-none"
    >
      Workspace: Delete active workspace
    </span>
    <span
      class="flex-1 flex "
    />
  </div>,
  <div
    class="flex side-bar-row flex-row items-center cursor-pointer   "
    style="padding-left: 16px; padding-right: 16px;"
  >
    <span
      class="text-lg truncate select-none"
    >
      Workspace: New workspace in Browser
    </span>
    <span
      class="flex-1 flex "
    />
  </div>,
  <div
    class="flex side-bar-row flex-row items-center cursor-pointer   "
    style="padding-left: 16px; padding-right: 16px;"
  >
    <span
      class="text-lg truncate select-none"
    >
      Workspace: New workspace in filesystem
    </span>
    <span
      class="flex-1 flex "
    />
  </div>,
]
`;

exports[`Correctly switches to file type 2`] = `
<div>
  <div>
    <div
      class="bangle-palette z-30 p-2 shadow-md border flex flex-col"
    >
      <div
        class="flex mb-2 sticky top-0"
      >
        <input
          aria-label="palette-input"
          class="flex-grow px-2"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex side-bar-row flex-row items-center cursor-pointer  active "
        style="padding-left: 16px; padding-right: 16px;"
      >
        <span
          class="text-lg truncate select-none"
        >
          one.md
        </span>
        <span
          class="flex-1 flex "
        />
      </div>
    </div>
  </div>
</div>
`;
